src/include/storage/index/in_mem_hash_index.h

```diff
@@ -80,21 +80,20 @@ class InMemHashIndex final {
     using Key = std::conditional_t<std::same_as<T, common::ku_string_t>, std::string_view, T>;
     // Appends the buffer to the index. Returns the number of values successfully inserted.
     // I.e. if a key fails to insert, its index will be the return value
-    size_t append(const IndexBuffer<BufferKeyType>& buffer, visible_func isVisible) {
-        reserve(indexHeader.numEntries + buffer.size());
+    size_t append(const IndexBuffer<BufferKeyType>& buffer, uint64_t bufferOffset,
+        visible_func isVisible) {
+        reserve(indexHeader.numEntries + buffer.size() - bufferOffset);
         // Do both searches after splitting. Returning early if the key already exists isn't a
         // particular concern and doing both after splitting allows the slotID to be reused
         common::hash_t hashes[BUFFER_SIZE];
-        for (size_t i = 0; i < buffer.size(); i++) {
+        for (size_t i = bufferOffset; i < buffer.size(); i++) {
             hashes[i] = HashIndexUtils::hash(buffer[i].first);
-        }
-        for (size_t i = 0; i < buffer.size(); i++) {
             auto& [key, value] = buffer[i];
             if (!appendInternal(key, value, hashes[i], isVisible)) {
-                return i;
+                return i - bufferOffset;
             }
         }
```

Comment:
A little late to comment on this, but if you're combining the loops then you may as well get rid of the hashes array and pass the results directly to appendInternal.

I'd seen similar results when I'd written it initially. It's worth noting that it doesn't help that we're targeting the default x86-64-v1 microarchitecture (from ~2003), so any any such tests should also try with newer instruction sets which could speed up auto-vectorized code (though I'd tried a while ago and didn't see a noticeable improvement yet from the newer micro-architectures; if we get to the point that there is then we might want to look at either targeting a newer version by default, or providing multiple builds).
I think it would require a significant amount of time to look into, but it might be beneficial at some point to try and get some vectorization working properly.

I have a suspicion that in the two loop version, the bulk of the CPU time is being spent in the second one, which isn't well suited to running multiple instructions in parallel (with all the branching), so in the combined version the hash calculations may actually be able to be done simultaneously with the more serial branching code (I don't know how much of this the CPU can pipeline in reality, but while the results of the hash function are needed immediately in appendInternal, the calculation of the next hash would be independent of the end of the previous call to appendInternal, and can just be discarded if the function has to return early). That's not to say there will be an improvement with the combined one; the pipelining may still slow things down slightly, but it's a hypothesis as to why there is little difference even assuming that the vectorized hashing is working properly in the two loop version.
This probably just isn't a very good place for vectorization.